lib/charms/operator_libs_linux/v2/snap.py

```diff
@@ -310,23 +310,35 @@ def _snap_daemons(
         except CalledProcessError as e:
             raise SnapError("Could not {} for snap [{}]: {}".format(args, self._name, e.stderr))
 
-    def get(self, key) -> str:
-        """Fetch a snap configuration value.
+    def get(self, key: str = "", *, typed: bool = False) -> Any:
+        """Fetch snap configuration values.
 
         Args:
-            key: the key to retrieve
+            key: the key to retrieve. Default to retrieve all values.
+            typed: set to True to retrieve typed values (set with typed=True).
+                Default is to return a string.
         """
+        if typed:
+            config = json.loads(self._snap("get", ["-d", key]))
+            if key:
+                return config.get(key)
+            return config
+
```

Comment:
I'm not comfortable with what this is doing when you call `get()` with `typed=False` -- the string returned is a table-formatted string (or perhaps empty with the way the code is currently written, as arg `''` will be sent?) which isn't very helpful.

So let's put this check here (in the `not typed` section):

```python
if not key:
    raise TypeError("key must be provided when typed=False")
```

Given that, I think `Optional[str]` would convey the type a bit better, but I don't have strong feelings about that.

Reply:
> I'm not comfortable with what this is doing when you call get() with typed=False -- the string returned is a table-formatted string

I think this is the current behaviour for nested config (although not called with `get()`):

```shell
root@sunbeam02:/tmp/lib/charms/operator_libs_linux/v2# snap get openstack-hypervisor compute
Key                          Value
compute.cpu-mode             host-model
compute.spice-proxy-address  10.6.0.134
compute.virt-type            kvm
```

```python>>> import snap
>>> s = snap.SnapCache()["openstack-hypervisor"]
>>> s.get("compute")
'Key                          Value\ncompute.cpu-mode             host-model\ncompute.spice-proxy-address  10.6.0.134\ncompute.virt-type
```

But I also agree that using `Optional[str]` will convey the meaning clearly. I will add the check and change it to `Optional[str]`

Reply:
Ah, that's interesting to know regarding nested config -- not very useful, but oh well. :-)

Thanks for the `raise TypeError` addition. I think that will at least prevent mistakes in common cases.